src/v/cluster/topic_table.cc

```diff
@@ -678,14 +679,16 @@ void topic_table::notify_waiters() {
 
     /// Consume all pending deltas
     if (!_waiters.empty()) {
-        changes.clear();
-        changes.swap(_pending_deltas);
-        _last_consumed_by_notifier_offset = 0;
         std::vector<std::unique_ptr<waiter>> active_waiters;
         active_waiters.swap(_waiters);
-        for (auto& w : active_waiters) {
-            w->promise.set_value(changes);
+        for (auto& w :
+             std::span{active_waiters.begin(), active_waiters.size() - 1}) {
+            w->promise.set_value(_pending_deltas);
         }
+
+        active_waiters[active_waiters.size() - 1]->promise.set_value(
+          std::move(_pending_deltas));
```

Comment:
i'm into the spirit of this optimization, but IIUC it only saves 1 out N copies at the expense of extra code and complexity, so the benefit isn't really clear.

if you decide to keep the optimization, though, you'll probably want to fix the final `std::move(_pending_deltas)` to explicitly express the desire to move and clear pending_deltas (ie something like std::exchange(_pending_deltas, {})) generally std::move(X) doesn't have the semantics that it guarantees anything is actually moved. For example, if `X` was `const` then `std::move(X)` will just do a copy.

Reply:
The main assumption here is that there tends to only be 1 waiter. The interface even warns against having multiple waiters. So in that sense we are eliminating all copies with this optimization. 

I will add a `_pending_deltas.clear()` after the move to just ensure the intent to clear it is obvious.

Reply:
maybe the deltas should be moved into a shared pointer and then change the interface so that the promises all get copies of a lw_shared_ptr? then it's 1 copy for any number of waiters?

Reply:
It may be better to just restrict the interface to only allow for one waiter. Beyond unit tests the `src/v/cluster/controller_backend.cc` is the only user of this interface currently. So it wouldn't be hard to change it to only allow for one waiter. And I don't believe the interface will ever work well with more than one waiter as unless the waiters all register at the same time none of them will get every pending delta.

Reply:
Baring that interface change though using a `shared_ptr` is definitely a great idea. Will switch the PR over to that if we decide not to change the interface.

Reply:
i wrote this with an assumption that there is only one waiter, now i would probably use `ss::queue` for this

Reply:
Will be making a separate PR to change to interface for `wait_for_changes`. This PR is related to an incident and needs to be merged asap. And I don't want to add any more complications to it.